src/roles/dullahan.py

```diff
@@ -103,100 +100,88 @@ def on_del_player(evt, cli, var, nick, nickrole, nicktpls, death_triggers):
+        evt.data["onlybywolves"].discard(d.nick)
+        evt.data["killers"][d.nick].append(k.nick)
 
 @event_listener("exchange_roles")
 def on_exchange(evt, cli, var, actor, nick, actor_role, nick_role):
-    if actor in KILLS:
-        del KILLS[actor]
-    if nick in KILLS:
-        del KILLS[nick]
-    if actor_role == "dullahan" and nick_role != "dullahan" and actor in TARGETS:
-        TARGETS[nick] = TARGETS[actor] - {nick}
-        del TARGETS[actor]
-    elif nick_role == "dullahan" and actor_role != "dullahan" and nick in TARGETS:
-        TARGETS[actor] = TARGETS[nick] - {actor}
-        del TARGETS[nick]
+    for k in set(KILLS):
+        if k.nick == actor or k.nick == nick:
+            del KILLS[k]
+
+    for k in set(TARGETS):
+        if actor_role == "dullahan" and nick_role != "dullahan" and k.nick == actor:
+            TARGETS[users._get(nick)] = TARGETS.pop(k) - {users._get(nick)} # FIXME
+        elif nick_role == "dullahan" and actor_role != "dullahan" and k.nick == nick:
+            TARGET[users._get(actor)] = TARGETS.pop(k) - {users._get(actor)} # FIXME
 
 @event_listener("chk_nightdone")
 def on_chk_nightdone(evt, cli, var):
-    spl = set(list_players())
+    spl = set(get_players())
     evt.data["actedcount"] += len(KILLS)
-    for p in var.ROLES["dullahan"]:
-        if TARGETS[p] & spl:
-            evt.data["nightroles"].append(p)
+    for d, targets in TARGETS.items():
+        if targets & spl:
+            evt.data["nightroles"].append(d.nick)
 
 @event_listener("transition_night_end", priority=2)
 def on_transition_night_end(evt, cli, var):
-    for dullahan in var.ROLES["dullahan"]:
-        targets = list(TARGETS[dullahan])
-        for target in var.DEAD:
-            if target in targets:
-                targets.remove(target)
-        if not targets: # already all dead
-            pm(cli, dullahan, "{0} {1}".format(messages["dullahan_simple"], messages["dullahan_targets_dead"]))
+    for dullahan, targets in TARGETS.items():
+        targs = list(targets)
+        for target in targets:
+            if target.nick in var.DEAD:
+                targs.remove(target) # FIXME: Update when var.DEAD holds User instances
+        if not targs: # already all dead
+            dullahan.send("{0} {1}".format(messages["dullahan_simple"], messages["dullahan_targets_dead"]))
             continue
-        random.shuffle(targets)
-        if dullahan in var.PLAYERS and not is_user_simple(dullahan):
-            pm(cli, dullahan, messages["dullahan_notify"])
-        else:
-            pm(cli, dullahan, messages["dullahan_simple"])
+        random.shuffle(targs)
+        dullahan.send(messages["dullahan_" + ("simple" if dullahan.prefers_simple() else "notify")])
         t = messages["dullahan_targets"] if var.FIRST_NIGHT else messages["dullahan_remaining_targets"]
-        pm(cli, dullahan, t + ", ".join(targets))
+        dullahan.send(t + ", ".join([t.nick for t in targs]))
```

Comment:
CPython implementation detail: If the iterable is not a list, it's implicitly converted to a list before performing the join. Yeah, it's an implementation detail, but I think that ultimately it doesn't matter.